### ISSUE TITLE
Warn on structs with a trailing zero-sized array that aren't repr(C)

### ISSUE THREAD

**Issue by Manishearth** [rust-lang/rust-clippy#2868]
```rust
struct Foo {
  /// ...
  last: [SomeType; 0]
}
```

such a struct is likely being created to pass to C (zero sized arrays aren't very useful in Rust itself) since the same pattern is sometimes used in C code for headers of allocations

credit @Gankro for the idea

**Comment by Gankro** [comment]
may also just be used in conjuction with manual allocation to make it easy to compute the offset of the array, but the result is the same; need repr(C).

**Comment by Gankro** [comment]
a leading array could also reasonably get a warning to use repr(align) or whatever instead.

**Comment by kennytm** [comment]
You can't use `#[repr(align)]` if the align-as type doesn't have a fixed size (e.g. `[usize; 0]` or even `[T; 0]`)

**Comment by brightly-salty** [comment]
I'm willing to attempt to implement this if it is still relevant.

**Comment by nhamovitz** [comment]
Hi o/ — first-time contributor here. I'm interested in trying to fix this. I've been reading some of the Clippy PRs lately so I have some idea where to start but I'll probably be back with questions 😅. 

@rustbot claim

**Comment by Manishearth** [comment]
Nah none of these should be linted imo

**Comment by nhamovitz** [comment]
When linting struct definitions, is it convention to include any attributes on the definition in the pretty-printing? i.e.
```
error: trailing zero-sized array in a struct which is not marked with a `repr` attribute
  --> $DIR/trailing_zero_sized_array_without_repr.rs:20:1
   |
LL | / #[must_use]
LL | | struct OnlyAnotherAttribute {
LL | |     field: i32,
LL | |     last: [usize; 0],
LL | | }
   | |_^
   |
   = help: consider annotating the struct definition with `#[repr(C)]` or another `repr` attribute
```

or would we start at the `struct` line?

**Comment by Manishearth** [comment]
Correct!
